zephyr/drivers/can/can_sccan.c

```diff
@@ -0,0 +1,1147 @@
+#define SCCAN_CRCER   BIT(8)
+#define SCCAN_RXFOVF  BIT(7)
+#define SCCAN_RXFUDF  BIT(6)
+#define SCCAN_RXFVAL  BIT(5)
+#define SCCAN_RCVDN   BIT(4)
+#define SCCAN_TXFOVF  BIT(3)
+#define SCCAN_TXHBOVF BIT(2)
+#define SCCAN_ARBLST  BIT(1)
+#define SCCAN_TRNSDN  BIT(0)
+
+/* CAN Interrupt Enable Register */
+#define SCCAN_IER_ALL_ENA (0x00003FFF)
+
+/* CAN TX Message Register1 */
+#define SCCAN_TXID1(x)    (x << 21)
+#define SCCAN_TXSRTR(x)   (x << 20)
+#define SCCAN_TXIDE(x)    (x << 19)
+#define SCCAN_TXID1(x)    (x << 21)
+#define SCCAN_TXID_EX1(x) ((x & GENMASK(28, 18)) << 3)
+#define SCCAN_TXID_EX2(x) ((x & GENMASK(17, 0)) << 1)
+#define SCCAN_TXERTR(x)   (x)
+
+/* CAN RX Message Register1 */
+#define SCCAN_RXID1_STD(x)        ((x & GENMASK(31, 21)) >> 21)
+#define SCCAN_RXID1_EXT(x)        ((x & GENMASK(31, 21)) >> 3)
+#define SCCAN_RX_STD_REMOTE_FRAME BIT(20)
+#define SCCAN_RX_EXT_FRAME        BIT(19)
+#define SCCAN_RXID2_EXT(x)        ((x & GENMASK(18, 1)) >> 1)
+#define SCCAN_RX_EXT_REMOTE_FRAME BIT(0)
+
+/* CAN RX Message Register2: */
+#define SCCAN_DLC(x) (x & GENMASK(3, 0))
+
+/* CAN Acceptance Filter ID Mask Register */
+#define SCCAN_AFID1(x)    (x << 21)
+#define SCCAN_AFSRTR(x)   (x << 20)
+#define SCCAN_AFIDE(x)    (x << 19)
+#define SCCAN_AFID1(x)    (x << 21)
+#define SCCAN_AFID_EX1(x) ((x & GENMASK(28, 18)) << 3)
+#define SCCAN_AFID_EX2(x) ((x & GENMASK(17, 0)) << 1)
+#define SCCAN_AFERTR(x)   (x)
+
+/* CAN FIFO and Buffer Reset Register */
+#define SCCAN_FIFORR_TXHPBRST  BIT(17)
+#define SCCAN_FIFORR_TXFIFORST BIT(16)
+#define SCCAN_FIFORR_RXFIFORST BIT(0)
+#define SCCAN_FIFO_CLEAR_ALL                                                                       \
+	(SCCAN_FIFORR_TXHPBRST | SCCAN_FIFORR_TXFIFORST | SCCAN_FIFORR_RXFIFORST)
+
+/* CAN Self Test Mode Control Register */
+#define SCCAN_STM_DISABLE (0U)
+#define SCCAN_STM_ENABLE  (1U)
+
+/* Timeout configuration for enable/disable CAN */
+#define SCCAN_ENABLE_RETRIES     (10U)
+#define SCCAN_ENABLE_DELAY_USEC  K_USEC(10)
+#define SCCAN_DISABLE_RETRIES    (10U)
+#define SCCAN_DISABLE_DELAY_MSEC K_MSEC(10)
+
+/* Timeout parameter for mutex_lock */
+#define SCCAN_MUTEX_LOCK_TIMEOUT K_MSEC(100)
+
+typedef void (*irq_init_func_t)(const struct device *dev);
+
+struct sc_can_cfg {
+	uint32_t reg_addr;
+	irq_init_func_t irq_init;
+	uint32_t clock_frequency;
+	uint32_t bus_speed;
+	uint8_t sjw;
+	uint16_t sample_point;
+	uint32_t max_bitrate;
+	uint8_t tx_fifo_depth;
+	int max_filter;
+};
+
+struct sc_can_tx_cb_data {
+	can_tx_callback_t tx_cb;
+	void *tx_cb_arg;
+};
+
+struct sc_can_rx_cb_data {
+	can_rx_callback_t rx_cb;
+	void *rx_cb_arg;
+};
+
+struct sc_can_data {
+	/*
+	 * This mutex protects:
+	 *  - Enabling/Disabling operations for CAN at the same time
+	 *  - Set CAN Timing operations (Need to set the multiple register)
+	 *  - Set TX parameter operations (Need to set the multiple parameter)
+	 *  - Set RX filter parameter operations (Need to set the multiple parameter)
```

Comment:
On second thoughts, enabling CAN by setting`CAN_EN` to `1` takes time because it will wait for 11 bits of recessive.  It's also take time to be `0` if it's still sending data.  Is it OK to return an error code when it's called while enabled.  IOW, can we make it a requirement to call `can_stop()` before calling `can_add_rx_fitler()`?

Reply:
> can we make it a requirement to call can_stop() before calling can_add_rx_fitler()?

Yes, I also want add/remove RX filter to be a precondition for can stop().
However, the current CAN sample add the RX filter after can_start().

https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/drivers/can/counter/src/main.c#L226-L249

In the current implementation, add/remove of RX filter automatically executes can_stop in this driver, so when the RX ISR happens, it is basically assumed that no one has touched the RX filter.